File: code/matlab/problem3b_1.m
% Overlay the problem 3(a)/3(b) beams plus the Problem 2 baseline.
% Lets us visually compare average vs worst-case designs and aperture size.

close all;
clear all;
clc;

%f=figure;


[x1, y1] = loadfromfig('problem3a.fig');
[x2, y2] = loadfromfig('problem3b.fig');
[x3, y3] = loadfromfig('problem2.fig');

xx = [x1 x2 x3];
yy = [y1 y2 y3];


f = figure;
plot(x1, y1, 'b')
hold on
plot(x2, y2, 'r')
plot(x3, y3, 'm')

grid on
xlim([-90 90])
ylim([-140, 0])

xlabel('Normal Angle (deg)')
ylabel('Array Response (dB)')

legend('20 antenna Epigraph (problem-3a)', ...
    '40 antenna Epigraph (problem-3b)', ...
    '40 antenna (problem-2a)', ...
    'Location', 'southeast')
hold off

saveas(f, 'problem3b_1.png')


function [x1_, y1_] = loadfromfig(filename)
    % Helper: open a .fig, read its line plot data, then close it again.
    f1 = openfig(filename);
    lineHandle = findobj(f1, 'Type', 'line');
    
    x1_ = get(lineHandle, 'XData');
    y1_ = get(lineHandle, 'YData');

    close(f1);
end

File: code/python/problem3b_1.py
"""Plot beam patterns from problem3a, problem3b, and problem2."""

import pickle
from pathlib import Path

import matplotlib.pyplot as plt


def load_curve(fig_path: Path):
    """Return x/y data from the first line of a stored Matplotlib figure."""
    with fig_path.open("rb") as fig_file:
        fig = pickle.load(fig_file)
    lines = [line for ax in fig.axes for line in ax.get_lines()]
    if not lines:
        raise ValueError(f"No line data in {fig_path}")
    xdata = lines[0].get_xdata()
    ydata = lines[0].get_ydata()
    plt.close(fig)
    return xdata, ydata


# (pickle filename, color, legend label)
fig_specs = [
    ("problem3a.fig.pickle", "b", "20 antenna Epigraph (problem-3a)"),
    ("problem3b.fig.pickle", "r", "40 antenna Epigraph (problem-3b)"),
    ("problem2.fig.pickle", "m", "40 antenna (problem-2a)"),
]

fig, ax = plt.subplots()
for path, color, label in fig_specs:
    x, y = load_curve(Path(path))
    ax.plot(x, y, color=color, label=label)

ax.grid(True)
ax.set_xlim([-90, 90])
ax.set_ylim([-140, 0])
ax.set_xlabel("Normal Angle (deg)")
ax.set_ylabel("Array Response (dB)")
ax.legend(loc="lower right")
fig.savefig("problem3b_1.png", dpi=300, bbox_inches="tight")
plt.show()
plt.close(fig)
